mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java

```diff
@@ -0,0 +1,279 @@
+            return Optional.empty();
+        }
+        if (!mailboxIdO.isPresent()) {
+            LOGGER.error("No mailbox id in message entry " + entry.getName());
+            return Optional.empty();
+        }
+        if (!sizeO.isPresent()) {
+            LOGGER.error("No size defined in message entry " + entry.getName());
+            return Optional.empty();
+        }
+        if (!internalDateO.isPresent()) {
+            LOGGER.error("No internal date defined in message entry " + entry.getName());
+            return Optional.empty();
+        }
+        if (!flagsO.isPresent()) {
+            LOGGER.error("No flags defined in message entry " + entry.getName());
+            return Optional.empty();
+        }
+
+        try {
+            return entryWithContent.getContent().map(Throwing.function(ByteSource::openStream).sneakyThrow())
+                .map(messageContentStream -> {
+                    MessageArchiveEntry messageArchiveEntry = new MessageArchiveEntry(messageIdO.get(),
+                        mailboxIdO.get(),
+                        sizeO.get(),
+                        internalDateO.get(),
+                        flagsO.get(),
+                        messageContentStream);
+                    return messageArchiveEntry;
+                });
+        } catch (Exception e) {
+            LOGGER.error("Error when opening input stream on content of message entry " + entry.getName(), e);
+            return Optional.empty();
+        }
+    }
+
+    private Optional<byte[]> getCurrentEntryContentBytes(ByteSource currentContent) {
+        try {
+            return Optional.ofNullable(currentContent.read());
+        } catch (IOException e) {
+            LOGGER.error("Error when reading entry content", e);
+            return Optional.empty();
+        }
+    }
+
+    private Optional<String> getAnnotationContent(ZipEntryWithContent entry) {
+        return entry.getContent()
+            .flatMap(content -> getCurrentEntryContentBytes(content))
+            .map(raw -> new String(raw, Charsets.UTF_8));
+    }
+
+    private MailboxAnnotationKey getAnnotationKey(MailboxWithAnnotationsArchiveEntry mailbox, ZipEntry entry) {
+        String key = entry.getName().substring((mailbox.getMailboxName() + "/" + Zipper.ANNOTATION_DIRECTORY + "/").length());
+        return new MailboxAnnotationKey(key);
+    }
+
+    private MailArchiveEntry from(ZipEntryWithContent current, ZipEntryType currentEntryType, Optional<ZipEntry> next) throws ZipException {
+        switch (currentEntryType) {
+            case MAILBOX:
+                return fromMailboxEntry(current, next);
+            case MAILBOX_ANNOTATION_DIR:
+                return next();
+            case MAILBOX_ANNOTATION:
+                return fromMailboxAnnotationEntry(current, next);
+            case MESSAGE:
+                Optional<MessageArchiveEntry> messageEntry = fromMessageEntry(current);
+                if (messageEntry.isPresent()) {
```

Comment:
can be:
```java
return fromMessageEntry(current)
    .orElse(new UnknownArchiveEntry(current.getEntryName()));
```

Reply:
it's not really good, please ignore it

Reply:
i have done it, it forced me to make the return type of fromMessageEntry more general.
As Optional are not covariant...
I used orElseGet instead of orElse to get a lazy behavior.

Reply:
What do you think of it ?